beacon-chain/rpc/beacon/validators_stream.go

```diff
@@ -0,0 +1,380 @@
+	epochDuration := time.Duration(params.BeaconConfig().SecondsPerSlot*params.BeaconConfig().SlotsPerEpoch) * time.Second
+	eth1Deposits := cache.New(epochDuration, epochDuration*2)
+	depositMutex := &sync.RWMutex{}
+	defer stateSub.Unsubscribe()
+
+	// Fetch our current epoch.
+	headState, err := bs.HeadFetcher.HeadState(bs.Ctx)
+	if err != nil {
+		return status.Error(codes.Internal, "Could not access head state")
+	}
+	if headState == nil {
+		return status.Error(codes.Internal, "Not ready to serve information")
+	}
+	currentEpoch := headState.Slot() / params.BeaconConfig().SlotsPerEpoch
+
+	// Handle messages from client.
+	go func() {
+		for {
+			msg, err := stream.Recv()
+			if err == io.EOF {
+				return
+			}
+			if err != nil {
+				// Errors handle elsewhere
+				select {
+				case <-stream.Context().Done():
+					return
+				case <-bs.Ctx.Done():
+					return
+				case <-stateSub.Err():
+					return
+				default:
+				}
+				log.WithError(err).Debug("Receive from validators stream listener failed; client probably closed connection")
+				return
+			}
+			switch msg.Action {
+			case ethpb.SetAction_ADD_VALIDATOR_KEYS:
+				pubKeysMutex.Lock()
+				// Create existence map to ensure we don't duplicate keys.
+				pubKeysMap := make(map[[48]byte]bool, len(msg.PublicKeys))
+				for _, pubKey := range pubKeys {
+					pubKeysMap[bytesutil.ToBytes48(pubKey)] = true
+				}
+				addedPubKeys := make([][]byte, 0, len(msg.PublicKeys))
+				for _, pubKey := range msg.PublicKeys {
+					if _, exists := pubKeysMap[bytesutil.ToBytes48(pubKey)]; !exists {
+						pubKeys = append(pubKeys, pubKey)
+						addedPubKeys = append(addedPubKeys, pubKey)
+					}
+				}
+				pubKeysMutex.Unlock()
+				// Send current info for the added public keys.
+				if validators, err := bs.generateValidatorInfo(bs.Ctx, addedPubKeys, eth1Deposits, depositMutex); err == nil {
+					for _, validator := range validators {
+						if err := stream.Send(validator); err != nil {
+							stream.Context().Done()
+							break
+						}
+					}
+				}
+			case ethpb.SetAction_REMOVE_VALIDATOR_KEYS:
+				msgPubKeysMap := make(map[[48]byte]bool, len(msg.PublicKeys))
+				for _, pubKey := range msg.PublicKeys {
+					msgPubKeysMap[bytesutil.ToBytes48(pubKey)] = true
+				}
+				pubKeysMutex.Lock()
+				max := len(pubKeys)
+				for i := 0; i < max; i++ {
+					if _, exists := msgPubKeysMap[bytesutil.ToBytes48(pubKeys[i])]; exists {
+						copy(pubKeys[i:], pubKeys[i+1:])
+						pubKeys = pubKeys[:len(pubKeys)-1]
+						i--
+						max--
+					}
+				}
+				pubKeysMutex.Unlock()
+			case ethpb.SetAction_SET_VALIDATOR_KEYS:
+				pubKeysMutex.Lock()
+				pubKeys = make([][]byte, 0, len(msg.PublicKeys))
+				for _, pubKey := range msg.PublicKeys {
+					pubKeys = append(pubKeys, pubKey)
+				}
+				pubKeysMutex.Unlock()
+				if validators, err := bs.generateValidatorInfo(bs.Ctx, msg.PublicKeys, eth1Deposits, depositMutex); err == nil {
```

Comment:
Why is the `SET_VALIDATOR_KEYS` query sending the info streams?

Reply:
Setting (or adding) the keys sends current data for the keys, then future updates as epochs tick past.